### PR TITLE
Datasets tutorial

### PR DESCRIPTION
**Pull request by Lalaland**
This adds a simple tutorial for the datasets module

### CODE REVIEW

File: tutorials/datasets_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+# This tutorial covers the two major dataset types in piton, EventCollection
+# PatientCollection
+
+target_directory = "dataset_tutorial_target"
+
+if os.path.exists(target_directory):
+    print("This demo expects an empty target directory")
+    exit(-1)
+else:
+    os.mkdir(target_directory)
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:18]:
```suggestion
import shutil
shutil.rmtree(target_directory)
os.makedirs(target_directory)
```
(annoying to remember to have to delete this directory each time the script is run)

File: tutorials/datasets_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        value=None,
+    ),
+]
+
+events = piton.datasets.EventCollection(
+    os.path.join(target_directory, "events")
+)
+
+# Once we create an events object we can just start writing
+with contextlib.closing(events.create_writer()) as writer:
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:51]: Can you explain what a `create_writer()` is in this context

File: tutorials/datasets_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    for event in dummy_events:
+        writer.add_event(patient_id=30, event=event)
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:53]: This feels like an important point -- that this appends events to an already existing file (so the user will need to explicitly delete the file if they want to start over)

File: tutorials/datasets_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+# We can also iterate over events
+
+with events.reader() as reader:
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:57]: What is a `reader` ?

File: tutorials/datasets_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+#####################################
+##### Part 2: Create patients #######
+#####################################
+
+# Piton stores patients within PatientCollections
+# These are simply generated right from EventCollections
+
+patients = events.to_patient_collection(
+    os.path.join(target_directory, "patients")
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:70]: Does the user need to specify the file name for each of these extracts (e.g. 'patients', 'events', etc.) -- I feel like it's cleaner to just have them specify a directory, and let Piton worry about the internals of that directory?

File: tutorials/datasets_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+# We can iterate over patients
+
+with patients.reader() as reader:
+    for patient in reader:
+        print(patient)
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:77]: Would print out `patient.events` here

File: tutorials/datasets_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+
+#############################################################
+##### Part 3: Apply a transformation to the  patients #######
+#############################################################
+
+# Piton allows you to perform transformations on patients in a straightforward manner
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:84]: What does 'transformation' mean?

File: tutorials/datasets_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def transform(input: piton.Patient) -> piton.Patient:
+    return piton.Patient(
+        patient_id=input.patient_id,
+        events=[
+            a for a in input.events if a.value != b"test_value"
```
Comment (Miking98) [inline tutorials/datasets_tutorial.py:91]: Explain why you need to have a bytes string here to interpret `a.value`

**Review by Miking98 — COMMENTED** [review]
I like this as a test script, but as a tutorial it is still a bit confusing.

I think converting this to an `.ipynb` would be helpful.

We could then add a few images (e.g. the three-step diagram we drew on the whiteboard going from OMOP -> EventCollection -> PatientCollection -> PatientDatabase) to help visualize what's happening + print out intermediate steps + have intermediate outputs be visible when someone is viewing this `.ipynb` online (i.e. they can see what's happening without having to download and run the script)

File: tutorials/datasets_tutorial.py (hunk 5 above)
Comment (Lalaland) [reply tutorials/datasets_tutorial.py:70]: This is allowing the users to specify a directory. The user can put any name in there.

**Comment by Lalaland** [comment]
> I think converting this to an .ipynb would be helpful.

Let's convert this to an ipynb only once we have unit tests. We really need this code to work and I want to keep it as is so we can test it simply right now

**Comment by jason-fries** [comment]
@Lalaland it sounds like you need a fast test script for debugging so let's fork since the goal of this PR is a tutorial. Unit tests are great, but the point of the tutorial is to sanity check the API, naming conventions, etc. @Miking98 made a lot of good points on clarity and such. So let's merge and create another branch for spinning up the tutorial.

**Review by jason-fries — APPROVED** [review]
Punting on major changes here so we can just merge as a test script and refocus on creating a separate tutorial for API refinement

**Pull request merged**